setuid-wrapper activation: Approximate atomicity

shlevy opened the pull request:
This makes the replacement of the old wapper dir with the new one
atomic if the kernel and FS support RENAME_EXCHANGE, and falls back to
at least ensuring the old wrapper dir remains on the FS if interrupted
during the (now smaller) inconsistent window.

Fixes #18124

shlevy commented:
@domenkozar please test, I don't have a working 16.09 system yet.

domenkozar commented on nixos/modules/security/setuid-wrappers.nix:
```diff
@@ -17,6 +17,12 @@ let
     '';
   };
 
+  setuid-swap - pkgs.runCommand "setuid-swap" {} ''
```
should be `s/-/=/`

shlevy replied:
Darn, too slow!

domenkozar commented on nixos/modules/security/setuid-swap.c:
```diff
@@ -0,0 +1,36 @@
+#define _GNU_SOURCE
```
(nitpick) Missing header comment explaining what this source does

shlevy replied:
Hmm, normally I'd agree, but IMO this is widespread enough not to need it. From https://www.gnu.org/software/libc/manual/html_node/Feature-Test-Macros.html#index-_005fGNU_005fSOURCE:
> If you define this macro, everything is included: ISO C89, ISO C99, POSIX.1, POSIX.2, BSD, SVID, X/Open, LFS, and GNU extensions. In the cases where POSIX.1 conflicts with BSD, the POSIX definitions take precedence.

In other words, "this is non-portable, give me all the extensions"

domenkozar replied:
No, I meant for the whole file, would be nice to have a description what these 36 lines do :)

shlevy replied:
Ah, OK, incoming

shlevy replied:
Pushed

shlevy committed:
setuid-wrapper activation: Approximate atomicity

This makes the replacement of the old wapper dir with the new one
atomic if the kernel and FS support RENAME_EXCHANGE, and falls back to
at least ensuring the old wrapper dir remains on the FS if interrupted
during the (now smaller) inconsistent window.

Fixes #18124

domenkozar commented:
I fear this will need a migration path also:

```
dummy> chmod: cannot access '/var/setuid-wrappers-new/polkit-agent-helper-1': No such file or directory
dummy> chown: cannot access '/var/setuid-wrappers-new/polkit-agent-helper-1': No such file or directory
dummy> chmod: cannot access '/var/setuid-wrappers-new/polkit-agent-helper-1': No such file or directory
dummy> cp: cannot create regular file '/var/setuid-wrappers-new/unix_chkpwd': No such file or directory
dummy> /nix/store/7jcj13r1a9r14436z6lmimxr0lhqi3lg-nixos-system-dummy-16.09.git.612b1a6/activate: line 269: /var/setuid-wrappers-new/unix_chkpwd.real: No such file or directory
dummy> chmod: cannot access '/var/setuid-wrappers-new/unix_chkpwd': No such file or directory
dummy> chown: cannot access '/var/setuid-wrappers-new/unix_chkpwd': No such file or directory
dummy> chmod: cannot access '/var/setuid-wrappers-new/unix_chkpwd': No such file or directory
dummy> swapping /var/setuid-wrappers-new and /var/setuid-wrappers: No such file or directory
```

domenkozar commented:
Oh `${wrapperDir}-new` needs to mkdir-ed first probably

shlevy commented:
Fixed

domenkozar commented:
`dummy> swapping /var/setuid-wrappers-new and /var/setuid-wrappers: Device or resource busy`

shlevy commented:
Hmm I wonder if that's because you're running under sudo and so one of the files is open? Anyway, pushed a fix.

domenkozar commented:
@shlevy I'm using what ever nixops is using.

Next error: `dummy> moving /var/setuid-wrappers-new to /var/setuid-wrappers-tmp: Directory not empty`

shlevy commented:
Hmm, what behavior do we want here, when there's an old tmp hanging around? Just clear it unconditionally? Keep one backup if it exists?

domenkozar commented:
@shlevy so that would happen only in case of failure during the activationScript?

I'd just error out and instruct to delete it manually since it's hard to know what happened.

edolstra commented:
Why not make `/var/setuid-wrappers` a symlink? Then it can be replaced atomically without requiring bleeding edge kernel features. Also, I've wanted to put `/var/setuid-wrappers` on a tmpfs for a long time anyway (so all other filesystems can be mounted `nosuid`), so we could just make it a symlink to somewhere on `/run` or some other tmpfs.

shlevy commented:
Originally I had the impression that setuid scripts living in a symlink dir was dangerous, but I misunderstood that issue. So will close this, go @edolstra's route

domenkozar commented:
@edolstra @shlevy so this was failing because it's ready a tmpfs upon install:

```
nixos/modules/installer/tools/nixos-install.sh:mount -t tmpfs -o "mode=0755" none $mountPoint/var/setuid-wrappers
```

So, if we want `/var/setuid-wrappers/` to be a tmpfs and a symlink that won't work.

My proposal would be to make `/var/setuid-wrappers/` a tmpfs and move conents to `/var/setuid-wrappers/bin` which could then be a symlink. This would be a backwards incompatible change, but we can fix it in nixpkgs and document it. 

Thoughts?
